### ISSUE TITLE
heap use after free in eval

### ISSUE THREAD

**Issue by Asmod4n** [mruby/mruby#3817]
Looks like eval has a heap use after free issue

```
>>> Test host <<<
mrbtest - Embeddable Ruby Test

mruby/src/vm.c:2369:11: runtime error: division by zero
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior mruby/src/vm.c:2369:11 in 
=================================================================
==5284==ERROR: AddressSanitizer: heap-use-after-free on address 0x60400017fae8 at pc 0x00010dbfba7f bp 0x7ffee27adbb0 sp 0x7ffee27ad350
READ of size 4 at 0x60400017fae8 thread T0
    #0 0x10dbfba7e in wrap_memcmp (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x1ca7e)
    #1 0x10d6027fe in str_eql string.c:974
    #2 0x10d60167a in mrb_str_equal string.c:991
    #3 0x10d621527 in mrb_str_equal_m string.c:1012
    #4 0x10d54f67c in mrb_funcall_with_block vm.c:473
    #5 0x10d549b07 in mrb_funcall_argv vm.c:490
    #6 0x10d54955f in mrb_funcall vm.c:374
    #7 0x10d487051 in mrb_equal object.c:48
    #8 0x10d4ecbd7 in mrb_obj_not_equal_m class.c:1671
    #9 0x10d5800d9 in mrb_vm_exec vm.c:1436
    #10 0x10d562ab9 in mrb_vm_run vm.c:922
    #11 0x10d5ebd75 in mrb_top_run vm.c:2912
    #12 0x10d6e8dfd in load_irep load.c:642
    #13 0x10d6e857e in mrb_load_irep_cxt load.c:648
    #14 0x10d6e908c in mrb_load_irep load.c:654
    #15 0x10d46578c in GENERATED_TMP_mrb_mruby_eval_gem_test (mrbtest:x86_64+0x10002778c)
    #16 0x10d445c37 in mrbgemtest_init mrbtest.c:108
    #17 0x10d4459c5 in mrb_init_mrbtest mrbtest.c:38
    #18 0x10d44473a in main driver.c:167
    #19 0x7fff6c77b144 in start (libdyld.dylib:x86_64+0x1144)

0x60400017fae8 is located 24 bytes inside of 48-byte region [0x60400017fad0,0x60400017fb00)
freed by thread T0 here:
    #0 0x10dc383e6 in wrap_free (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x593e6)
    #1 0x10d54006b in mrb_default_allocf state.c:56
    #2 0x10d7181bd in mrb_free gc.c:272
    #3 0x10d5411a1 in mrb_irep_free state.c:165
    #4 0x10d540706 in mrb_irep_decref state.c:133
    #5 0x10d71ac35 in obj_free gc.c:828
    #6 0x10d72c098 in incremental_sweep_phase gc.c:1064
    #7 0x10d728961 in incremental_gc gc.c:1130
    #8 0x10d721d16 in incremental_gc_until gc.c:1146
    #9 0x10d71fe8b in mrb_incremental_gc gc.c:1197
    #10 0x10d71f035 in mrb_obj_alloc gc.c:514
    #11 0x10d5ef7dd in mrb_str_new_capa string.c:102
    #12 0x10d52119d in join_ary array.c:1108
    #13 0x10d51fc05 in mrb_ary_join array.c:1156
    #14 0x10d52ffe3 in mrb_ary_join_m array.c:1176
    #15 0x10d5800d9 in mrb_vm_exec vm.c:1436
    #16 0x10d562ab9 in mrb_vm_run vm.c:922
    #17 0x10d5ebd75 in mrb_top_run vm.c:2912
    #18 0x10d6e8dfd in load_irep load.c:642
    #19 0x10d6e857e in mrb_load_irep_cxt load.c:648
    #20 0x10d6e908c in mrb_load_irep load.c:654
    #21 0x10d46578c in GENERATED_TMP_mrb_mruby_eval_gem_test (mrbtest:x86_64+0x10002778c)
    #22 0x10d445c37 in mrbgemtest_init mrbtest.c:108
    #23 0x10d4459c5 in mrb_init_mrbtest mrbtest.c:38
    #24 0x10d44473a in main driver.c:167
    #25 0x7fff6c77b144 in start (libdyld.dylib:x86_64+0x1144)

previously allocated by thread T0 here:
    #0 0x10dc385b0 in wrap_realloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x595b0)
    #1 0x10d540085 in mrb_default_allocf state.c:60
    #2 0x10d71622c in mrb_realloc_simple gc.c:204
    #3 0x10d7170ba in mrb_realloc gc.c:218
    #4 0x10d717f23 in mrb_malloc gc.c:240
    #5 0x10d542a47 in mrb_str_pool state.c:201
    #6 0x10d7950e0 in new_lit codegen.c:492
    #7 0x10d75e620 in codegen codegen.c:2339
    #8 0x10d78bc14 in gen_values codegen.c:854
    #9 0x10d748144 in codegen codegen.c:1653
    #10 0x10d73c947 in codegen codegen.c:1287
    #11 0x10d7764aa in scope_body codegen.c:742
    #12 0x10d7469c8 in codegen codegen.c:1597
    #13 0x10d736013 in mrb_generate_code codegen.c:3011
    #14 0x10d8f2827 in create_proc_from_string eval.c:247
    #15 0x10d8ef567 in f_eval eval.c:301
    #16 0x10d5800d9 in mrb_vm_exec vm.c:1436
    #17 0x10d562ab9 in mrb_vm_run vm.c:922
    #18 0x10d5ebd75 in mrb_top_run vm.c:2912
    #19 0x10d6e8dfd in load_irep load.c:642
    #20 0x10d6e857e in mrb_load_irep_cxt load.c:648
    #21 0x10d6e908c in mrb_load_irep load.c:654
    #22 0x10d46578c in GENERATED_TMP_mrb_mruby_eval_gem_test (mrbtest:x86_64+0x10002778c)
    #23 0x10d445c37 in mrbgemtest_init mrbtest.c:108
    #24 0x10d4459c5 in mrb_init_mrbtest mrbtest.c:38
    #25 0x10d44473a in main driver.c:167
    #26 0x7fff6c77b144 in start (libdyld.dylib:x86_64+0x1144)

SUMMARY: AddressSanitizer: heap-use-after-free (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x1ca7e) in wrap_memcmp
Shadow bytes around the buggy address:
  0x1c080002ff00: fa fa 00 00 00 00 00 fa fa fa 00 00 00 00 00 fa
  0x1c080002ff10: fa fa 00 00 00 00 00 fa fa fa 00 00 00 00 00 fa
  0x1c080002ff20: fa fa fd fd fd fd fd fa fa fa 00 00 00 00 00 fa
  0x1c080002ff30: fa fa 00 00 00 00 00 fa fa fa fd fd fd fd fd fa
  0x1c080002ff40: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fa
=>0x1c080002ff50: fa fa 00 00 00 00 00 00 fa fa fd fd fd[fd]fd fd
  0x1c080002ff60: fa fa fd fd fd fd fd fd fa fa fd fd fd fd fd fa
  0x1c080002ff70: fa fa 00 00 00 00 00 00 fa fa fd fd fd fd fd fd
  0x1c080002ff80: fa fa 00 00 00 00 00 07 fa fa fd fd fd fd fd fd
  0x1c080002ff90: fa fa 00 00 00 00 00 07 fa fa fa fa fa fa fa fa
  0x1c080002ffa0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==5284==ABORTING
.................................................................................................................................................................................?...............................................................................................................................................................................................rake aborted!
Command failed with status (): ["build/host/bin/mrbtest"...]
mruby/lib/mruby/build.rb:299:in `run_test'
mruby/Rakefile:123:in `block (2 levels) in <top (required)>'
mruby/lib/mruby/build.rb:13:in `instance_eval'
mruby/lib/mruby/build.rb:13:in `block in each_target'
mruby/lib/mruby/build.rb:12:in `each'
mruby/lib/mruby/build.rb:12:in `each_target'
mruby/Rakefile:122:in `block in <top (required)>'
Tasks: TOP => test
(See full trace by running task with --trace)
```

**Comment by Asmod4n** [comment]
@matz any update on this?

**Comment by matz** [comment]
Probably it's caused by #3829 which I'm working on.

**Comment by matz** [comment]
@Asmod4n I think I fixed this along with #3829 but if you see the issue unfixed, please reopen.
